### DISCUSSION TITLE
ESP32 Serial.setDebugOutput

### DISCUSSION THREAD

**Discussion by Xylopyrographer** [espressif/arduino-esp32#7639] [Q&A]
For the ESP32 can someone shed light on `Serial.setDebugOutput()`?

Only info I can find is from the ESP8266 docs on Serial [here](https://arduino-esp8266.readthedocs.io/en/latest/reference.html), where it states:

> By default the diagnostic output from WiFi libraries is disabled when you call Serial.begin. To enable debug output again, call `Serial.setDebugOutput(true)`. To redirect debug output to Serial1 instead, call `Serial1.setDebugOutput(true)`.
> 
> You also need to use `Serial.setDebugOutput(true)` to enable output from printf() function.

I see the function is in the current core libraries (`HardwareSerial.h` for example), and I can trace it through to the lower level libraries but does it affect anything?

 Is this lingering code from porting from the ESP8266 to the ESP32? 

`.print()`, `.println()`, `.printf()`, `log_x()`, and `LOG_X()` all work OK without  ever using `.setDebugOutput( true )` ( even with using WiFi).

Is it turned on by default somewhere in the 'Serial' libraries?

**Comment by mrengineer7777** [comment]
In my experience .setDebugOutput(true) doesn't seem to do anything.

**Reply by me-no-dev** [reply]
It actually does, but I think it's ON for Serial by default. If you want to have the output on some other port (Serial1, USB, etc) then it will be switched. We should have it off by default though, so maybe @SuGlider can have a look

**Comment by SuGlider** [comment]
Sorry for the long delay in answering it. For the records:

`HardwareSerial::setDebugOutPut(bool)` sets the Serial port (UARTx) to output DEBUG information using the Arduino `log_x()` where `x` is `d` for debug, `e` for error or `i` for information log messages.

By default, right after booting, Logging is set to output using UART0 (console serial port). 
If the user wants to mute all `log_x()` output, `Serial.setDebugOutput(false)` will do the job.
`Serial.end()` also does the same, among other things, such as detaching UART pins.

By other hand `Serial1.setDebugOutput(true)` would set all the log messages to be sent to UART1.
Only 1 serial port can output log messages. The last Serial port that set `DebugOutput` to `true` will be used.

Let's say that I want to use UART0 for connecting to an external device, but I also want to see the logging into UART1. 
The code for that shall be:
``` cpp
Serial.begin(9600);  // connected to an external UART hardware
Serial1.begin(115200);  // Our log messages port
Serial1.setDebugOutput(true);  // enables logging to be sent to UART1 and not to UART0 anymore
log_i("This log message goes to Serial1!");
```
### Details:

`setDebugOutPut(true)` sets a function pointer in ROM ETS PUTC1, which in turn is used by  ROM `ets_printf()`, which is used by IDF and Arduino log functions. Therefore, it just points the `ets_install_putc1()` to a void function that writes a character to the UARTx FIFO in order to transmit it through the Serial Port.
